src/expr/skolem_manager.cpp

```diff
@@ -59,25 +49,31 @@ Node SkolemManager::mkSkolem(Node v,
                              ProofGenerator* pg,
                              bool retWitness)
 {
+  // We do not currently insist that pred does not contain witness terms
   Assert(v.getKind() == BOUND_VARIABLE);
   // make the witness term
   NodeManager* nm = NodeManager::currentNM();
   Node bvl = nm->mkNode(BOUND_VAR_LIST, v);
-  // translate pred to witness form, since pred itself may contain skolem
-  Node predw = getWitnessForm(pred);
-  // make the witness term, which should not contain any skolem
-  Node w = nm->mkNode(WITNESS, bvl, predw);
+  // Make the witness term, where notice that pred may contain skolem. We do
+  // not recursively convert pred to witness form, since witness terms should
+  // be treated as opaque. Moreover, the use of witness forms leads to
```

Comment:
What is the impact of this in terms of `getWitnessForm`? If one has a skolem and wants to get its witness form without skolems it's their responsibility to do the recursive transformation?

Reply:
Yes, which notice would be quite complicated, it would involve variable shadowing issues.

Reply:
I see. It'd be good to add a note on that for getWitnessForm documentation, saying that since it's not recursive one can get witness forms with skolems and there is no guarantee that the witnesses of these skolems can be substituted as is without leading to shadowing.